### ISSUE TITLE
Characterize and emulate 'total nonsense' in yaAGC

### ISSUE THREAD

**Issue by thewonderidiot** [virtualagc/virtualagc#1047]
@indy91 and I found an interesting snippet of code in Zerlina's P66 while working through the scalings to use for all of its padloaded values:
```
EXTEND                                  # IS |VDGVERT - VVECTX| < VERCRIT?
DCA             MPAC
EXTEND
DV              VBUF            +1
EXTEND
BZF             ERRFAIL
```
Leading into this, (VDGVERT - VVECTX) is in MPAC,MPAC+1, and VERCRIT is in VBUF+1. The purpose here is to decide, based on the magnitude of (VDGVERT - VVECTX), whether to use TAURODB or TAURODL for the ROD computation.

This is an interesting sort of test that I haven't seen before, and it looks like it might run into trouble in yaAGC. If (VDGVERT - VVECTX) is >= VERCRIT here, then we run into a situation where the dividend is greater than the divisor. According to Savage&Drake, this results in "total nonsense", and yaAGC presently returns random numbers in both A and L. However, if I am reading this code correctly, it looks like Don is assuming (and testing that) A receives +0 or -0 in such situations.

We haven't yet verified whether or not this is actually happening, but if such tests are done it would be a good idea to figure out exactly what 'total nonsense' entails and properly emulate it.

**Comment by rburkey2005** [comment]
As usual, I didn't try to deduce what the result would be when I handled this case, and if I'm not mistaken, literally return a random value using a random number generator.  :smile:  So I wouldn't be surprised if you're right, and if this isn't another mystery for you to verify the correct behavior of.

**Comment by thewonderidiot** [comment]
Well, we just ran a test with a lightly modified version of Zerlina that saved off the value (VDGVERT - VVECTX) for inspection. The typical value for that was 00000,XXXXX or 77777,XXXXX, getting sometimes up to 00004,XXXXX if multiple ROD clicks are done at once. The value currently being used for VERCRIT is 11272, so the division is quite safe, contrary to my initial belief. So luckily we don't need to handle nonsense for Zerlina to work. :)

I think I am still going to figure out what goes on in such cases, and add support for it to yaAGC and some sample cases to Borealis, just for accuracy's sake (and to nip any future questions like this in the bud).

**Comment by rburkey2005** [comment]
The name of the game is simulation accuracy, even if I was relatively far away from achieving it by myself.  I've often thought that in providing high-level descriptions of the instructions and peripheral devices (which nevertheless weren't quite complete or correct), and then supplementing those descriptions with lower-level descriptions of the microcode (which were better but still not perfect), and then supplementing _those_ with the expectation of consulting the electrical schematics, the original AGC project used a very curious approach to creating a high-reliability system.  Or else it was a result of the population of computer engineers being of a far different nature than the population of computer engineers we have today.  

Imagine how that would work out today:  "Here's the README and tutorial for node.js," we'd tell new programmers, "and then if you have any questions, here are the schematics of a PC."  :smile:

**Comment by thewonderidiot** [comment]
Haha! :smile: So, I've been playing with some numbers tonight, and to be fair to Savage and Drake, so far it sure seems like what I'm getting is pretty garbage. I feel like there is a tiny bit of a pattern going on.... but I haven't quite placed my finger on it yet. I think I am going to write a C version of what the hardware actually does, so I can use that instead of the hardware sim to get expected results.

It might be easiest to then just stick that hardware model into yaAGC, but I fear it might diminish performance a bit... the structure of it is going to look something like:
```
<short setup calculations>
for (int i = 0; i < 14; i++)
{
    <conditional calculations>
}
<short wrap-up calculations>
```

But maybe it will be fast enough to not matter. There is something nice about the standard case for DV as currently implemented using actual divide and modulo operators.

**Comment by rburkey2005** [comment]
If it is used only for the one case you mention, which never comes up in real life, it's hard to see why that would affect performance.

**Comment by thewonderidiot** [comment]
Good point!

I haven't fully validated it yet, and it is by no means optimized, but here is a really basic version of what the registers are doing (with just a couple of shortcuts taken, so the numbers printed the last time through the loop aren't quite correct, but the final results are). It's given me good numbers for everything I've tried so far, including correct total nonsense results:
```C
#include <stdio.h>
#include <stdint.h>

#define AGC_P1 ((int16_t) 1)

int
AddSP16 (int Addend1, int Addend2)
{
  int Sum;
  Sum = Addend1 + Addend2;
  if (Sum & 0200000)
    {
      Sum += AGC_P1;
      Sum &= 0177777;
    }
  return (Sum);
}

int main(int argc, char **argv)
{
    uint16_t a;
    uint16_t l;
    uint16_t d;

    uint16_t qsign = 0;
    uint16_t dsign = 0;
    uint16_t b = 0;
    uint16_t lsign = 0;
    uint16_t bsign = 0;
    uint16_t g = 0;
    uint16_t u = 0;
    uint16_t y = 0;

    if (argc != 4)
    {
        return 1;
    }
    sscanf(argv[1], "%ho", &a);
    sscanf(argv[2], "%ho", &l);
    sscanf(argv[3], "%ho", &d);

    printf("Start: %06o,%06o / %06o\n", a, l, d);

    // DV0.1 RA WB TSGN
    qsign = a & 0100000;

    // DV0.2[0X]
    if (!qsign)
    {
        a = ~a;
    }

    // DV0.1, DV0.2[0X] TMZ
    if (a == 0177777)
    {
        // DV1.4[X1] TSGN
        // DV1.5 Z16
        qsign = l & 0100000;
    }

    // DV1.4 RL WB
    if (qsign)
    {
        // DV1.5 RC
        l = ~l;
    }

    // DV1.5 RB/RC WY B15X
    // DV1.6 RU WL
    l = AddSP16(l, 040000);

    printf("DV1,6: A=%06o, L=%06o\n", a, l);

    // DV1.6 TOV
    if ((l & 0140000) == 0040000)
    {
        //DV1.8[X1] RA WY
        //DV1.10 RU WB
        b = a;
    }
    else
    {
        //DV1.8[X1] RA WY PONEX
        //DV1.10 RU WB
        b = AddSP16(a, 1);
    }

    // DV1.7 RG RSC WB TSGN
    // DV1.9[1X] Z15
    dsign = d & 0100000;
    if (!dsign)
    {
        // DV1.0[0X] RB WA
        a = d;
    }
    else
    {
        // DV1.0[0X] RC WA
        a = ~d;
    }

    // DV1.11 RL WYD
    // DV1.12 RU WL
    lsign = l >> 15;
    l = (lsign << 15) | ((l & 037777) << 1) | lsign;
    printf("DV1,12: A=%06o, L=%06o, B=%06o\n", a, l, b);

    for (int i = 0; i < 14; i++)
    {
        // L.1 L2GD
        g = (lsign << 15) | ((l & 037777) << 1);
        // L.1 RB WYD
        bsign = b >> 15;
        y = (bsign << 15) | ((b & 037777) << 1);
        // L.1 PIFL
        if ((l & 040000) == 0)
        {
            // L.1: WYD
            y |= bsign;
        }
        // L.1 A2X
        u = AddSP16(y, a);
        // L.2 RG WL
        l = g;
        lsign = l >> 15;
        // Slightly wrong for last iteration... but we can recover
        // L.2 TSGU
        if (u & 0100000)
        {
            // L.2[1X] RB1F
            // L.3 RU WB
            l |= 1;
            b = u;
        }
        else
        {
            // L.2[0X] CLXC
            // L.3 RU WB
            b = y;
        }
        printf("Loop %2d: G=%06o, L=%06o, B=%06o\n", i, g, l, b);
    }
    // DV4.5 RG WB WA
    a = l;
    // DV4.6 RZ TOV
    if (qsign != dsign)
    {
        // DV4.7 RC WA
        a = ~l;
    }
    
    // DV4.9 RU WB WL
    l = b;
    // DV4.8 TZ TSGN
    if (!qsign)
    {
        // DV4.10[0X] RC WL
        l = ~l;
    }
    printf("Results: A=%06o, L=%06o\n", a, l);
}
```

**Comment by indy91** [comment]
Looks like a lot of code for a case that never happens.

**Comment by rburkey2005** [comment]
@indy91 True.  But since the effect on performance and the size of the executables is insignificant, the amount of effort put into the solution is only important before the work has been done ... i.e., in deciding whether to put in the effort or not.  Since @thewonderidiot has already put in the work, mostly, the cost is now zero!  :smile:  Plus, my vote is for getting as close to 100% accuracy in the simulation as humanly possible.

**Comment by indy91** [comment]
I'm just speaking from laziness, haha, I'll have to manually update our yaAGC files with this DV change. I'm of course also for 100% accuracy.

**Comment by rburkey2005** [comment]
I can certainly sympathize with that.  Never do anything which isn't necessary; very zen.  :smile:  My understanding, though, is that even though it's on the hasn't-been-done-yet list, @thewonderidiot 's  intention is to try to unify those two code bases to make those updates much easier.  I hadn't really realized there was a problem until I saw issue #622.  I recall making some changes many years ago to facilitate the process of importing Virtual AGC code into NASSP, and hadn't realized there were continuing difficulties with that.  It would certainly be best if the code was structured in such a way as to make it usable to you as-is.

**Comment by thewonderidiot** [comment]
Yeah, it's a bit longer than I expected. :smile: But really, I wrote this because it is _way way_ easier to analyze cases using this model than it is to extract the relevant numbers from the waveforms of the hardware sim. I suspect this could be streamlined a lot.... and I would like to come up with a more compelling explanation of what the numbers actually are than just "total nonsense" for the assembly language manual page.

**Comment by thewonderidiot** [comment]
Right, so I haven't reduced it to a formula yet, but I have looked at a lot of nonsense divisions today and there is definitely a rather simple pattern behind it. I'm able to predict some of the nonsense results now. The basic patten I'm seeing is this:

Given a division  of A,L / D, the results will be:
* A = 37777 - x
* L = A + y (not overflow corrected)

Here, y counts up based on the initial value of A,L, starting from 0 (if L was 0) up until A + y == 77777 - D. One A,L value higher, and the resulting L wraps to 0 while x increments by 1. So for example,

10001,00000 / 10001 = 37777, 10001 (as per the documentation)
10001,00001 / 10001 = 37777, 10002
10001,37777 / 10001 = 37777, 50000
10002,00000 / 10001 = 37777, 50001
10002,17775 / 10001 = 37777, 67776
10002,17776 / 10001 = 37776, 00000

This continues for a bit, and wraps can happen multiple (up to many) times. Eventually, though, the whole thing wraps back to 37777, 00000. But I haven't identified exactly when that is yet, or what happens from there. But hopefully tomorrow I'll be able to come up swith some much shorter nonsense- generating code. 😄

**Comment by rburkey2005** [comment]
In the future you may be known as the Edward Lear of the AGC.

**Comment by thewonderidiot** [comment]
Well, maybe not. :P There's (perhaps unsurprisingly) more going on in this than I originally thought. So, I think I am instead just going to clean up the above code and plug it straight in for the nominal cases.

I noticed, by the way, that A, L, and Z are all (potentially) modified before "RSC RG WB" happens in timepulse 7 of DV1. "RSC WB" will be the pulses responsible for reading the divisor, if the divisor is in the central registers. So `DV A`, `DV L`, and `DV Z` will all potentially give unexpected results, none of which yaAGC currently emulates. So I'll try to make sure the right thing happens there too.

**Comment by thewonderidiot** [comment]
To expand on those unexpected results:
* `DV A` -- when you do this, what you actually divide by is -|A|. 
* `DV Z` -- this divides by Z with bit 16 set -- meaning your divisor gets a sudden negative overflow condition, and the division will proceed as if it were negative.
* `DV L` -- this is a bit more complicated. If the quotient is negative, per DV sign rules, L is inverted. Then (regardless of signs) 40000 is added to it. This result will be the divisor.

**Comment by thewonderidiot** [comment]
Progress update: I've shortened/simplified/made human-readable the above code and integrated it into yaAGC. It is now producing correct nonsense results for every combination I've tried.

I've also put in special handling for A, L, and Z. I haven't yet tested L. I've tested A and it seems to be functional. Z works when A,L is positive, but when A,L is negative, it gets the answers completely wrong. It's making me wonder if overflow in the divisor drastically changes the results of division... I need to dig into it more.
